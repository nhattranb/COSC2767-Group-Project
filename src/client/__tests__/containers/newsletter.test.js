/*
 *   RMIT University Vietnam
 *   Course: COSC2767 Systems Deployment and Operations
 *   Semester: 2024C
 *   Assessment: Assignment 2
 *   Author: Le Dinh Ngoc Quynh
 *   ID: s3791159
 *   Created  date: 25 Dec 2024
 *   Last modified: 25 Dec 2024
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Newsletter from '../../app/containers/Newsletter';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

// Mock actions
jest.mock('../../app/actions', () => ({
  newsletterChange: jest.fn(),
  subscribeToNewsletter: jest.fn()
}));

describe('Newsletter component', () => {
  let store;

  const initialState = {
    newsletter: {
      email: '',
      isSubscribed: false
    }
  };

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  it('renders correctly', () => {
    const stateWithFormErrors = {
      ...initialState,
      newsletter: {
        ...initialState.newsletter,
        formErrors: {}
      }
    };
    store = mockStore(stateWithFormErrors);

    const { asFragment } = render(
      <Provider store={store}>
        <Newsletter />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders subscribed state correctly', () => {
    const subscribedState = {
      newsletter: {
        email: 'test@example.com',
        isSubscribed: true
      }
    };
    store = mockStore(subscribedState);

    const subscribedStateWithFormErrors = {
      ...subscribedState,
      newsletter: {
        ...subscribedState.newsletter,
        formErrors: {}
      }
    };
    store = mockStore(subscribedStateWithFormErrors);

    const { asFragment } = render(
      <Provider store={store}>
        <Newsletter />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
}); 