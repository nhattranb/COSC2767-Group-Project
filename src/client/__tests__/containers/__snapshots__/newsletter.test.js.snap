// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletter component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="newsletter-form"
  >
    <p>
      Subscribe to Our Newsletter
    </p>
    <form>
      <div
        class="subscribe"
      >
        <div
          class="input-box inline-btn-box "
        >
          <div
            class="input-text-block"
          >
            <input
              autocomplete="on"
              class="input-text"
              name="email"
              placeholder="Enter Your Email Address"
              type="text"
              value=""
            />
            <button
              class="input-btn custom-btn-primary md text-only icon-left "
              style="border-radius: 3px;"
              type="submit"
            >
              <span
                class="btn-text"
              >
                Join Now
              </span>
            </button>
          </div>
          <span
            class="invalid-message"
          />
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Newsletter component renders subscribed state correctly 1`] = `
<DocumentFragment>
  <div
    class="newsletter-form"
  >
    <p>
      Subscribe to Our Newsletter
    </p>
    <form>
      <div
        class="subscribe"
      >
        <div
          class="input-box inline-btn-box "
        >
          <div
            class="input-text-block"
          >
            <input
              autocomplete="on"
              class="input-text"
              name="email"
              placeholder="Enter Your Email Address"
              type="text"
              value="test@example.com"
            />
            <button
              class="input-btn custom-btn-primary md text-only icon-left "
              style="border-radius: 3px;"
              type="submit"
            >
              <span
                class="btn-text"
              >
                Join Now
              </span>
            </button>
          </div>
          <span
            class="invalid-message"
          />
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
